File: Graphs/prims.c
/* File Name: prims.c */

#include <stdio.h>
#include <limits.h>

#define MAX 100

int minKey(int key[], int mstSet[], int vertices) {
    int min = INT_MAX, min_index;

    for (int v = 0; v < vertices; v++)
        if (mstSet[v] == 0 && key[v] < min) {
            min = key[v];
            min_index = v;
        }

    return min_index;
}

void primMST(int graph[MAX][MAX], int vertices) {
    int parent[MAX];
    int key[MAX];
    int mstSet[MAX];

    for (int i = 0; i < vertices; i++) {
        key[i] = INT_MAX;
        mstSet[i] = 0;
    }

    key[0] = 0;       // Start from vertex 0
    parent[0] = -1;   // Root of MST

    for (int count = 0; count < vertices - 1; count++) {
        int u = minKey(key, mstSet, vertices);
        mstSet[u] = 1;

        for (int v = 0; v < vertices; v++)
            if (graph[u][v] && mstSet[v] == 0 && graph[u][v] < key[v]) {
                parent[v] = u;
                key[v] = graph[u][v];
            }
    }

    printf("Edge \tWeight\n");
    for (int i = 1; i < vertices; i++)
        printf("%d - %d \t%d\n", parent[i], i, graph[i][parent[i]]);
}

int main() {
    int vertices;
    int graph[MAX][MAX];

    printf("Enter number of vertices: ");
    scanf("%d", &vertices);

    printf("Enter weighted adjacency matrix:\n");
    for (int i = 0; i < vertices; i++)
        for (int j = 0; j < vertices; j++)
            scanf("%d", &graph[i][j]);

    primMST(graph, vertices);

    return 0;
}

File: Graphs/prims.py
# File Name: prims.py

import sys

def min_key(key, mst_set, vertices):
    min_val = sys.maxsize
    min_index = -1

    for v in range(vertices):
        if not mst_set[v] and key[v] < min_val:
            min_val = key[v]
            min_index = v

    return min_index

def prim_mst(graph, vertices):
    parent = [-1] * vertices
    key = [sys.maxsize] * vertices
    mst_set = [False] * vertices

    key[0] = 0

    for _ in range(vertices - 1):
        u = min_key(key, mst_set, vertices)
        mst_set[u] = True

        for v in range(vertices):
            if graph[u][v] != 0 and not mst_set[v] and graph[u][v] < key[v]:
                parent[v] = u
                key[v] = graph[u][v]

    print("Edge \tWeight")
    for i in range(1, vertices):
        print(f"{parent[i]} - {i}\t{graph[i][parent[i]]}")

vertices = int(input("Enter number of vertices: "))
print("Enter weighted adjacency matrix:")

graph = []
for _ in range(vertices):
    graph.append(list(map(int, input().split())))

prim_mst(graph, vertices)
